language-plutus-core/src/Language/PlutusCore/Universe/Core.hs

```diff
@@ -0,0 +1,348 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DefaultSignatures     #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE KindSignatures        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE TemplateHaskell       #-}
+
+module Language.PlutusCore.Universe.Core
+    ( Some (..)
+    , In (..)
+    , Of (..)
+    , SomeIn
+    , SomeOf
+    , someValue
+    , Includes (..)
+    , Closed (..)
+    , type (<:)
+    , knownUniOf
+    , GShow (..)
+    , gshow
+    , GEq (..)
+    , (:~:) (..)
+    , GLift (..)
+    ) where
+
+import           Control.DeepSeq
+import           Control.Monad
+import           Data.GADT.Compare
+import           Data.GADT.Show
+import           Data.Hashable
+import           Data.Proxy
+import           Data.Text.Prettyprint.Doc (Pretty (..))
+import           GHC.Exts
+import           Language.Haskell.TH.Lift
+import           Language.Haskell.TH.Syntax
+import           Text.Show.Deriving
+
+{- Note [Universes]
+A universe is a collection of types. It can be finite like
+
+    data U a where
+        UUnit :: U ()
+        UInt  :: U Int
+
+or infinite like
+
+    data U a where
+        UBool :: U Bool
+        UList :: !(U a) -> U [a]
+
+Here are some values of the latter 'U' / the types that they encode
+
+        UBool               / Bool
+        UList UBool         / [Bool]
+        UList (UList UBool) / [[Bool]]
+
+'U' being a GADT allows to package a type from a universe together with a value of that type.
+For example,
+
+    Some (Of UBool True) :: Some (Of U)
+-}
+
+-- | Existential quantification as a data type.
+data Some f = forall a. Some (f a)
+
+-- | A particular type from a universe.
+newtype In uni a = In (uni a)
+
+-- | A value of a particular type from a universe.
+data Of uni a = Of (uni a) a
+
+-- | Some type from a universe.
+type SomeIn uni = Some (In uni)
+
+-- | A value of some type from a universe.
+type SomeOf uni = Some (Of uni)
+
+-- We probably want to use that together with `fastsum`.
+-- | A constraint for \"@a@ is in @uni@\".
+class uni `Includes` a where
+    knownUni :: uni a
```

Comment:
Which method? `fastsum`? For creating universes on the fly, we talked about that, will elaborate in a comment.

Reply:
Actually, I decided to delete that comment as `fastsum` will only work for finite universes and it's not really clear if we want to create universes on the fly at all (certainly not right now).

Reply:
I meant `knownUni`.

Reply:
Added to Note [Universes]:

> We say that a type is in a universe whenever there is a tag for that type in the universe.
For example, 'Int' is in 'U', because there exists a tag for 'Int' in 'U' -- 'UInt'.